Firestore/core/src/firebase/firestore/remote/watch_stream.h

```diff
@@ -41,20 +43,49 @@ namespace firebase {
 namespace firestore {
 namespace remote {
 
+/**
+ * An interface defining the events that can be emitted by the `WatchStream`.
+ */
+class WatchStreamCallback {
+ public:
+  /** Called by the `WatchStream` when it is ready to accept outbound request
+   * messages. */
+  virtual void OnWatchStreamOpen() = 0;
+
+  /**
+   * Called by the `WatchStream` with changes and the snapshot versions
+   * included in in the `WatchChange` responses sent back by the server.
+   */
+  virtual void OnWatchStreamChange(
+      const WatchChange& change,
+      const model::SnapshotVersion& snapshot_version) = 0;
+
+  /**
+   * Called by the `WatchStream` when the underlying streaming RPC is
+   * interrupted for whatever reason, usually because of an error, but possibly
+   * due to an idle timeout. The status passed to this method may be ok, in
+   * which case the stream was closed without attributable fault.
+   *
+   * NOTE: This will not be called after `Stop` is called on the stream. See
+   * "Starting and Stopping" on `Stream` for details.
+   */
+  virtual void OnWatchStreamClose(const util::Status& status) = 0;
+};
+
 /**
  * A `Stream` that implements the StreamingWatch RPC.
  *
- * Once the `WatchStream` has called the `streamDidOpen` method on the delegate,
- * any number of `WatchQuery` and `UnwatchTargetId` calls can be sent to control
- * what changes will be sent from the server for WatchChanges.
+ * Once the `WatchStream` has called the `OnWatchStreamOpen` method on the
+ * callback, any number of `WatchQuery` and `UnwatchTargetId` calls can be sent
+ * to control what changes will be sent from the server for WatchChanges.
  */
 class WatchStream : public Stream {
  public:
   WatchStream(util::AsyncQueue* async_queue,
               auth::CredentialsProvider* credentials_provider,
               FSTSerializerBeta* serializer,
               GrpcConnection* grpc_connection,
-              id<FSTWatchStreamDelegate> delegate);
+              WatchStreamCallback* callback);
```

Comment:
Consider adding something like the following on the docstring:
```
@param callback Called when open/change/etc events occur on this WatchStream. Must not be null[*]. Must remain valid for the lifetime of this object.

[*] - Or 'Allowed to be null, in which case, the callbacks won't be invoked'.
```

(I really only care about the middle sentence; you could omit the other two, as they seem sufficiently obvious.)

Due to obj-c's nil dispatch, you'd get the [*] behaviour for free. We should decide whether we explicitly want this or not.